Add Custom Dictionaries to Dockerized Spell Checker

RajKiranAcharyya opened the pull request:
### **PR: Add Custom Dictionaries to Dockerized Spell Checker**  

#### **Description:**  
This PR improves the **Dockerized spell checker** by adding support for **custom dictionaries** and ensuring proper volume mounting.  

#### **Key Changes:**  
✅ Added `cspell.json` with dictionary configurations.  
✅ Mounted `/custom-dicts/` in Docker for extra dictionaries.  
✅ Ensured `.avi`, `.dif`, and other dictionary files are recognized.  
✅ Fixed dictionary loading issues inside the container.  

### **Steps For Verification:**  
- [ ] Ensure that **custom dictionaries are correctly loaded** when running `cspell` inside the container.  
- [ ] Run `ls -lah /app/custom-dicts` inside the container to confirm that dictionary files are accessible.  
- [ ] Validate `cspell --config /app/cspell.json /app/test.txt` runs without dictionary errors.  

#### **Next Steps & Feedback Needed:**  
- Verify if dictionaries are correctly loaded inside the container.  
- Should we dynamically load dictionaries instead of static references in `cspell.json`?  

Looking forward to feedback! 🚀

arkid15r reviewed (changes requested):
I don't see any related changes in this PR.

Please make sure to double-check any work you submit next time. Thank you!

KJhellico commented:
@RajKiranAcharyya,
Let me just ask you about your vision: how this wonderful set of files should work?

RajKiranAcharyya commented:
> @RajKiranAcharyya,
> Let me just ask you about your vision: how this wonderful set of files should work?


Hi KJhellico, thanks for this question!
My vision is to integrate custom dictionaries dynamically, ensuring scalability for new languages and terms without rebuilding the container. Let me know if you have any suggestions!

KJhellico commented:
And...? How can we see this practically?

RajKiranAcharyya commented:
> And...? How can we see this practically?

Hi @KJhellico ,

I hope you’re doing well. I wanted to add a quick note that I should have included in the PR. Here’s a brief summary of the changes:

**Containerization: The Dockerized spell checker mounts /custom-dicts, allowing updates to dictionaries defined in cspell.json to be applied without rebuilding the image.

Verification:

List the dictionary files:

powershell
docker run --rm -it my-spellchecker ls -lah /app/custom-dicts

Run a spell check on your test file:

powershell
docker run --rm -v D:\avc\holidays\custom-dicts:/app/custom-dicts my-spellchecker --config /app/cspell.json /app/test.txt**


Please note that the full set of dictionaries for all languages is not included at this time—only some in cspell.json are mounted. I would appreciate your feedback on whether this approach is acceptable as a starting point before I continue with further updates.

Thank you, and I look forward to your thoughts!

Best regards, Raj

RajKiranAcharyya commented:
> > And...? How can we see this practically?
> 
> Hi @KJhellico ,
> 
> I hope you’re doing well. I wanted to add a quick note that I should have included in the PR. Here’s a brief summary of the changes:
> 
> **Containerization: The Dockerized spell checker mounts /custom-dicts, allowing updates to dictionaries defined in cspell.json to be applied without rebuilding the image.
> 
> Verification:
> 
> List the dictionary files:
> 
> powershell docker run --rm -it my-spellchecker ls -lah /app/custom-dicts
> 
> Run a spell check on your test file:
> 
> powershell docker run --rm -v D:\avc\holidays\custom-dicts:/app/custom-dicts my-spellchecker --config /app/cspell.json /app/test.txt**
> 
> Please note that the full set of dictionaries for all languages is not included at this time—only some in cspell.json are mounted. I would appreciate your feedback on whether this approach is acceptable as a starting point before I continue with further updates.
> 
> Thank you, and I look forward to your thoughts!
> 
> Best regards, Raj

KJhellico commented:
Requirements from #2308:
- Integration with pre-commit for automated spell checking.
- Compatibility with CI/CD pipelines (GitHub Actions, etc.).

RajKiranAcharyya commented:
Hi @arkid15r   and @KJhellico  ,

I appreciate the feedback and understand why the PR was closed.

- I have worked on integrating pre-commit hooks and ensuring CI/CD compatibility as required in #2308.
- Currently, the full set of dictionaries for all languages isn't included, but I am exploring ways to expand support.

As a beginner, I'm still learning and facing some challenges with these integrations. Your feedback, guidance, and the demo repository have been incredibly helpful. I'll work on addressing these issues and making the necessary improvements.

Please let me know if there are any additional adjustments you'd like me to prioritize. Thank you!

Best regards,  
Raj





**ci-cd_UPDATE.yml**

name: CI/CD

on:
  merge_group:
  pull_request:
    branches:
      - dev
      - main
      - v1
  push:
    branches:
      - dev
      - main
      - v1
    tags:
      - 'v*'
  workflow_dispatch:
  workflow_run:
    workflows: ['Update pre-commit hooks']
    branches:
      - update-pre-commit-hooks
    types:
      - completed

env:
  FORCE_COLOR: 1

concurrency:
  cancel-in-progress: ${{ !contains(fromJSON('["dev", "main", "v1"]'), github.ref_name) }}
  group: ${{ github.repository }}-${{ github.workflow }}-${{ github.head_ref || github.ref_name }}

jobs:
  pre-commit:
    name: Run pre-commit
    runs-on: ubuntu-24.04
    steps:
      - name: Check out repository
        uses: actions/checkout@11bd71901bbe5b1630ceea73d27597364c9af683

      - name: Set up Python
        uses: actions/setup-python@8d9ed9ac5c53483de85588cdf95a591a75ab9f55
        with:
          check-latest: true
          python-version: '3.13'

      - name: Run pre-commit
        uses: pre-commit/action@2c7b3805fd2a0fd8c1884dcaebf91fc102a13ecd

  code-ql:
    name: CodeQL
    needs:
      - pre-commit
    permissions:
      security-events: write
    runs-on: ubuntu-24.04
    steps:
      - name: Check out repository
        uses: actions/checkout@11bd71901bbe5b1630ceea73d27597364c9af683

      - name: Initialize CodeQL
        uses: github/codeql-action/init@1b549b9259bda1cb5ddde3b41741a82a2d15a841
        with:
          languages: python

      - name: Perform CodeQL Analysis
        uses: github/codeql-action/analyze@1b549b9259bda1cb5ddde3b41741a82a2d15a841
        with:
          category: '/language:python'

  test:
    name: Test Python ${{ matrix.python-version }} on ${{ matrix.os }}
    runs-on: ${{ matrix.os }}
    needs:
      - pre-commit
    strategy:
      matrix:
        os:
          - macos-latest
          - ubuntu-latest
          - windows-latest
        python-version:
          - '3.9'
          - '3.10'
          - '3.11'
          - '3.12'
          - '3.13'
    steps:
      - name: Check out repository
        uses: actions/checkout@11bd71901bbe5b1630ceea73d27597364c9af683

      - name: Set up Python ${{ matrix.python-version }}
        uses: actions/setup-python@8d9ed9ac5c53483de85588cdf95a591a75ab9f55
        with:
          allow-prereleases: true
          cache: pip
          cache-dependency-path: |
            requirements/runtime.txt
            requirements/tests.txt
          check-latest: true
          python-version: ${{ matrix.python-version }}

      - name: Install dependencies
        run: |
          python -m pip install --upgrade pip
          python -m pip install --requirement requirements/tests.txt
          python -m pip install .

      - name: Run tests
        run: |
          make test

      - name: Upload coverage to Codecov
        uses: codecov/codecov-action@0565863a31f2c772f9f0395002a31e3f06189574
        with:
          token: ${{ secrets.CODECOV_TOKEN }}

  build:
    name: Build distribution
    needs: test
    runs-on: ubuntu-24.04
    steps:
      - name: Check out repository
        uses: actions/checkout@11bd71901bbe5b1630ceea73d27597364c9af683

      - name: Set up Python
        uses: actions/setup-python@8d9ed9ac5c53483de85588cdf95a591a75ab9f55
        with:
          cache: pip
          cache-dependency-path: |
            requirements/build.txt
          check-latest: true
          python-version: '3.12'

      - name: Install dependencies
        run: |
          python -m pip install --upgrade pip
          python -m pip install -r requirements/build.txt
          python -m pip install .

      - name: Build distribution
        run: |
          make package

      - name: Upload package artifacts
        uses: actions/upload-artifact@ea165f8d65b6e75b540449e92b4886f43607fa02
        with:
          name: dist
          path: dist

      - name: Set version
        if: startsWith(github.event.ref, 'refs/tags/v')
        run: echo "VERSION=$(echo ${{ github.ref_name }} | sed 's/^v//')" >> $GITHUB_ENV

      - name: Generate SBOM
        if: startsWith(github.event.ref, 'refs/tags/v')
        run: |
          make sbom > holidays-${{ env.VERSION }}-sbom.json

      - name: Upload SBOM
        if: startsWith(github.event.ref, 'refs/tags/v')
        uses: actions/upload-artifact@ea165f8d65b6e75b540449e92b4886f43607fa02
        with:
          name: sbom
          path: holidays-${{ env.VERSION }}-sbom.json

  test-build:
    name: Test build on ${{ matrix.os }}
    runs-on: ${{ matrix.os }}
    needs: build
    strategy:
      matrix:
        os:
          - macos-latest
          - ubuntu-latest
          - windows-latest
    steps:
      - name: Check out repository
        uses: actions/checkout@11bd71901bbe5b1630ceea73d27597364c9af683

      - name: Set up Python
        uses: actions/setup-python@8d9ed9ac5c53483de85588cdf95a591a75ab9f55
        with:
          cache: pip
          cache-dependency-path: |
            requirements/runtime.txt
            requirements/tests.txt
          check-latest: true
          python-version: '3.13'

      - name: Get package artifacts
        uses: actions/download-artifact@95815c38cf2ff2164869cbab79da8d1f422bc89e
        with:
          name: dist
          path: dist

      - name: Run tests
        shell: bash
        run: |
          rm -rf holidays
          python -m pip install --requirement requirements/tests.txt
          python -m pip install `ls dist/*.whl`
          pytest --dist loadscope --numprocesses auto tests/countries tests/financial
          python -m pip uninstall -y holidays python-dateutil six
          python -m pip install `ls dist/*.tar.gz`
          pytest --dist loadscope --numprocesses auto tests/countries tests/financial

  test-docs:
    name: Test docs build
    runs-on: ubuntu-24.04
    needs: test
    steps:
      - name: Check out repository
        uses: actions/checkout@11bd71901bbe5b1630ceea73d27597364c9af683

      - name: Set Up Python
        uses: actions/setup-python@8d9ed9ac5c53483de85588cdf95a591a75ab9f55
        with:
          cache: pip
          cache-dependency-path: requirements/docs.txt
          python-version: '3.13'

      - name: Install dependencies
        run: |
          python -m pip install --requirement requirements/docs.txt
          python -m pip install .

      - name: Build docs
        run: |
          make doc

  docker-spell-check:
    name: Docker Spell Checker Test
    needs: pre-commit
    runs-on: ubuntu-latest
    steps:
      - name: Check out repository
        uses: actions/checkout@11bd71901bbe5b1630ceea73d27597364c9af683

      - name: Build Docker Image
        run: docker build --no-cache -t my-spellchecker .

      - name: Verify /custom-dicts Directory
        run: docker run --rm my-spellchecker ls -lah /app/custom-dicts

      - name: Run Spell Check Test
        run: |
          docker run --rm -v "${{ github.workspace }}/custom-dicts:/app/custom-dicts" my-spellchecker --config /app/cspell.json /app/test.txt

  publish-main:
    name: Publish generated artifacts
    if: |
      github.repository == 'vacanza/holidays' &&
      github.event_name == 'push' &&
      startsWith(github.event.ref, 'refs/tags/v')
    environment: main
    needs:
      - test-build
      - test-docs
    permissions:
      contents: write
      id-token: write
    runs-on: ubuntu-24.04
    steps:
      - name: Download package artifacts
        uses: actions/download-artifact@95815c38cf2ff2164869cbab79da8d1f422bc89e
        with:
          name: dist
          path: dist

      - name: Publish package distributions to PyPI
        uses: pypa/gh-action-pypi-publish@76f52bc884231f62b9a034ebfe128415bbaabdfc

  sign-artifacts:
    name: Create SHA1 checksums and Sigstore signatures
    runs-on: ubuntu-24.04
    needs:
      - publish-main
    permissions:
      id-token: write
    steps:
      - name: Download package artifacts
        uses: actions/download-artifact@95815c38cf2ff2164869cbab79da8d1f422bc89e
        with:
          name: dist
          path: dist

      - name: Compute SHA1 checksums
        run: |
          cd dist
          for file in *; do
            sha1sum "$file" > "$file.sha1"
          done

      - name: Sign the files using Sigstore
        uses: sigstore/gh-action-sigstore-python@f514d46b907ebcd5bedc05145c03b69c1edd8b46
        with:
          inputs: |
            ./dist/*.tar.gz
            ./dist/*.whl

      - name: Upload package dist and signatures
        uses: actions/upload-artifact@ea165f8d65b6e75b540449e92b4886f43607fa02
        with:
          name: signed-artifacts
          path: dist

  update-github-release:
    name: Update GitHub release with SBOM and signed artifacts
    runs-on: ubuntu-24.04
    needs:
      - sign-artifacts
    permissions:
      contents: write
    steps:
      - name: Download SBOM
        uses: actions/download-artifact@95815c38cf2ff2164869cbab79da8d1f422bc89e
        with:
          name: sbom

      - name: Download package dist and signatures
        uses: actions/download-artifact@95815c38cf2ff2164869cbab79da8d1f422bc89e
        with:
          name: signed-artifacts
          path: dist

      - name: Update Github release
        env:
          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
        run: |
          gh release upload --repo vacanza/holidays ${{ github.ref_name }} dist/*
          gh release upload --repo vacanza/holidays ${{ github.ref_name }} holidays-*-sbom.json






**pre-commit-autoupdate.yml**

name: Update pre-commit hooks

on:
  schedule:
    - cron: '0 0 * * *'
  workflow_dispatch:

env:
  FORCE_COLOR: 1

jobs:
  auto-update:
    name: Update pre-commit hooks
    if: ${{ github.repository == 'vacanza/holidays' }}
    permissions:
      contents: write
      pull-requests: write
    runs-on: ubuntu-24.04
    steps:
      - uses: actions/checkout@11bd71901bbe5b1630ceea73d27597364c9af683

      - uses: actions/setup-python@8d9ed9ac5c53483de85588cdf95a591a75ab9f55
        with:
          cache: pip
          python-version: '3.12'

      - name: Install dependencies
        run: |
          python -m pip install pre-commit

      - name: Use pre-commit environment cache
        uses: actions/cache@5a3ec84eff668545956fd18022155c47e93e2684
        with:
          key: ${{ runner.os }}-pre-commit-${{ hashFiles('.pre-commit-config.yaml') }}
          path: ~/.cache/pre-commit
          restore-keys: ${{ runner.os }}-pre-commit-

      - name: Run pre-commit automatic update
        run: |
          pre-commit autoupdate
          pre-commit run --all-files

      - uses: peter-evans/create-pull-request@271a8d0340265f705b14b6d32b9829c1cb33d45e
        with:
          base: dev
          body: Update pre-commit hooks to their latest versions.
          branch: update-pre-commit-hooks
          commit-message: 'chore: Update pre-commit hooks'
          committer: github-actions[bot] <github-actions[bot]@users.noreply.github.com>
          delete-branch: true
          title: Update pre-commit hooks
          token: ${{ github.token }}

arkid15r commented:
I suggest focusing on more beginner-friendly task (see the [Todo](https://github.com/orgs/vacanza/projects/3) section)
